<div style="float: left; border: 1px; width:33%; text-align: justify" >
    We analyze our sequence and variant data using a modified version of the <a href="https://software.broadinstitute.org/gatk/best-practices/">Broad Institute / GATK SNP and Indel Discovery Best Practices</a>, adapted for macaques.
    The major difficulty for accurate variant calling in species outside of human is the lack of high-confidence 'truth' datasets. To mitigate this, we developed a fairly strict quality filtering approach, validated using Mendelian violations and other analyses to identify regions of the macaque genome with systematic errors in short read alignment or genotype calling.
</div>
<div style="float: left;margin-left: 30px;">
    <img width="700" src="<%=contextPath%>/mgap/images/variantCalling.png" alt="Data Processing / Variant Calling Strategy">
</div>
